keps/sig-node/4603-tune-crashloopbackoff/README.md

```diff
@@ -0,0 +1,1221 @@
+rolling updates, and enhanced workload types like StatefulSets assume this,
+while other workload types like those implemented with the Job and CronJob API
+better model workloads that do exit themselves, running until Success or at
+predictable intervals. The end user choice not to use one of these alternative
+workload types, but to run a relatively fast exiting Pod (under 10 minutes) with
+both a successful exit code and configured to `restartPolicy: Always`, is
+interpreted now directly as the intention for the end user to restart the pod
+quickly and without penalty to resume/restart its workloads without
+rescheduling.
+
+TODO: pic!
+
+This provides a workaround (and therefore, opportunity for abuse), where
+application developers could catch any number of internal errors of their
+workload in their application code, but exit with exit code 0, forcing extra
+fast restart behavior in a way that is opaque to kubelet or the cluster
+operator. Something similar is already being taken advantage of by application
+developers via wrapper scripts, but this causes no direct extra strain on
+kubelet as it simply causes the container to run indefinitely.
+
+
+
+**Alternative**: Workloads must opt-in with `restartPolicy: FastOnSuccess`, as a
+foil to `restartPolicy: OnFailure`. In this case, existing workloads with
+`restartPolicy: Always` or ones not determined to be in the critical path would
+use the new, yet still relatively slower, front-loaded decay curve and only
+those updated with `FastOnSuccess` would get truer fast restart behavior.
+However, then it becomes impossible for a workload to opt into both
+`restartPolicy: FastOnSuccess` and `restartPolicy: Rapid`.
+
+#### API opt in for max cap decay curve (`restartPolicy: Rapid`)
+
+For some users in
+[Kubernetes#57291](https://github.com/kubernetes/kubernetes/issues/57291), any
+delay over 1 minute at any point is just too slow, even if it is legitimately
+crashing. A common refrain is that for independently recoverable errors,
+especially system infrastructure events or recovered external dependencies, or
+for absolutely nonnegotiably critical sidecar pods, users would rather poll more
+often or more intelligently to reduce the amount of time a workload has to wait
+to try again after a failure. In the extreme cases, users want to be able to
+configure (by container, node, or exit code) the backoff to close to 0 seconds.
+This KEP considers it out of scope to implement fully user-customizable
+behavior, and too risky to node stability to allow legitimately crashing
+workloads to have a backoff of 0, but it is in scope to provide users a way to
+opt workloads in to a faster restart curve that is not as drastic as what is
+intended for `Success` states, nor as beholden to the status quo as the new
+default front loaded decay with interval modification.
+
+Pods and init/sidecar containers will be able to set a new OneOf value,
```

Comment:
Thanks, you're right, and I've updated this to reflect that this only applies to sidecar aka "restartable init containers" for now since that is properly in line with the original intention here that `Rapid` is only a variation on `Always` for all purposes, including setting restart containers.

That being said I recognize your larger point still stands on separating the concerns so that the term is not overloaded, especially for pure init containers which otherwise cannot represent this since they by definition do not have a `restartPolicy`. Interested in more thoughts on this, as part of the impetus behind that decision was to minimize API changes.

Reply:
If the pod restart policy is "Rapid", I assume that would still affect init containers on failure?

Reply:
If `restartPolicy: Rapid` is a variant of `restartPolicy: Always` for the pod-level restart policy, does this mean that the backoff behavior on Init containers failure will differ between pods created by Deployments or StatefulSets and pods created by Jobs?

Reply:
> If the pod restart policy is "Rapid", I assume that would still affect init containers on failure?

I think this would make sense if I gave people a way to opt into it, except that by design init containers (NOT sidecar containers) can only have `restartPolicy: Never`, so that is why this would not be an option with the current API solution

> does this mean that the backoff behavior on Init containers failure will differ between pods created by Deployments or StatefulSets and pods created by Jobs?

No, but only because as above I'm not actually providing a way for init containers to use `restartPolicy: Rapid` with the current API direction.

FYI, I'm noodling with exposing this as a separate field in an alpha2 or giving initContainers a faster hardcoded curve directly as a result of this convo.

Reply:
@toVersus Somewhat related, this API decision is not immediately compatible with [KEP-3329: Retriable and non-retriable Pod failures for Jobs](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures) in a way that I think is a bit surprising, as that feature depends on `restartPolicy: Never` and then does the restarts on its own. See the new section in README.md#relationship-with-job-api-podfailurepolicy-and-backofflimit for some analysis on that which I think you might be interested in

Reply:
Thanks for sharing! Indeed, this behavior is surprising. I understand their reasoning, but I feel strange with the idea of abusing the meaning of `restartPolicy: Never` to use the custom implementation of the Pod restart mechanism in the Job controller. Since this is a custom implementation by the Job controller, it's probably fine to leave it outside the scope of this KEP. If the WG-Batch wants to change the backoff behavior in the Job controller, they will likely address this by adding new field to the Job API 😄

Reply:
FYI also added a note about the new issue in 299d699ebe132c09944de3c4b5d1c2adf8c24d7a